torch/csrc/distributed/c10d/reducer.cpp

```diff
@@ -2052,6 +2052,34 @@ void verify_params_across_processes(
     const c10::intrusive_ptr<c10d::ProcessGroup>& process_group,
     const std::vector<at::Tensor>& params,
     const c10::optional<std::weak_ptr<c10d::Logger>>& logger) {
+
+  // First verify number of parameters to avoid inconsistent inputs into
+  // broadcast which can cause a crash.
+  // See https://github.com/pytorch/pytorch/issues/73547
+
+  at::TensorOptions param_size_options;
+  param_size_options = param_size_options.dtype(at::kLong);
+  param_size_options = param_size_options.device(params[0].device());
+  // Note: Not using tensor building API because of
+  // https://github.com/pytorch/pytorch/issues/74114
+  at::Tensor param_size_tensor = at::tensor(
+    {static_cast<int64_t>(params.size())}, param_size_options);
+
+  // Broadcast and verify parameter size.
+  std::vector<at::Tensor> param_size_vec{param_size_tensor};
+  process_group->broadcast(param_size_vec)->wait();
```

Comment:
nit (can be done as future BE work item): DDP ctor will broadcast param values anyway. Is it possible to implement these two broadcasts in the same location using two async ops? I recall the param value broadcast was done in the `distributed.py` file?

Reply:
Yes, there's actually 3 collectives that go on now:

1. parameter size verification
2. parameter shape verification
3. parameter broadcast

We could make these all async, and wait on all three, and register then() callbacks on the first two to raise appropriate errors. Filed https://github.com/pytorch/pytorch/issues/74185